doc/designs/quic-design/json-encoder.md

```diff
@@ -0,0 +1,91 @@
+JSON Encoder
+============
+
+Approach
+--------
+
+The JSON encoder exists to support QLOG implementation. There is no intention to
+implement a decoder at this time. The encoder is intended to support
+zero-allocation automation using immediate calls without the use of an
+intermediate syntax tree representation. This enables highly efficient
+serialization when called from QUIC code without dynamic memory allocation.
+
+An example usage is as follows:
+
+```c
+int generate_json(BIO *b)
+{
+    int ret = 1;
+    JSON_ENC z;
+
+    if (!ossl_json_init(&z, b, 0))
+        return 0;
+
+    ossl_json_object_begin(&z);
+    {
+        ossl_json_key(&z, "key");
+        ossl_json_str(&z, "value");
+
+        ossl_json_key(&z, "key2");
+        ossl_json_u64(&z, 42);
+
+        ossl_json_key(&z, "key3");
+        ossl_json_array_begin(&z);
+        {
+            ossl_json_null(&z);
+            ossl_json_f64(&z, 42.0);
+            ossl_json_str(&z, "string");
+        }
+        ossl_json_array_end(&z);
+    }
+    ossl_json_object_end(&z);
+
+    if (ossl_json_get_error_flag(&z))
+        ret = 0;
+
+    ossl_json_cleanup(&z);
+    return ret;
+}
+```
+
+The zero-allocation, immediate-output design means that most API calls
```

Comment:
It can happen, though generally won't. Since we have a fixed stack of 128 bits before we have to spill out to using a heap-allocated stack. This allows 128 levels of nesting composite structures which would be quite extreme for QLOG. So the JSON encoder avoids the need for allocations in likely cases, but will function correctly if it does happen.